clippy_lints/src/operators/arithmetic_side_effects.rs

```diff
@@ -134,14 +129,18 @@ impl ArithmeticSideEffects {
         ) {
             return;
         };
-        if self.is_allowed_ty(cx, lhs) || self.is_allowed_ty(cx, rhs) {
+        if self.is_allowed_ty(cx, lhs) && self.is_allowed_ty(cx, rhs) {
```

Comment:
That does leave us in the position where e.g. `Complex<usize> * usize` requires allowing `Complex` and `usize`, which would also allow all `usize` operations. Not exactly sure what the best thing to do about that is

Reply:
Well, it is not possible to conclude anything in regards to custom types or how they interactive with other things, therefore, it seems plausible to `&&` instead of `II`.

The matter of allowing heterogeneous combinations could be addressed in a future PR asking for users to specify pairs instead of single values in `arithmetic-side-effects-allowed`.

```
arithmetic-side-effects-allowed  = [
  "Complex<usize>,usize",
  "Complex<f64>,*",
  "*,Complex<f64>"
]
```

Reply:
Specifying pairs seems reasonable, in that case I'd suggest adding a check that the `lhs` and `rhs` types are the same for now

Reply:
Yeah just until the pairs thing is implemented

Reply:
Done as requested